source/darwin/nw_socket.c

```diff
@@ -0,0 +1,2232 @@
+    } else {
+        s_process_socket_cancel_task(&args->task, args, AWS_TASK_STATUS_RUN_READY);
+    }
+}
+
+static void s_socket_internal_destroy(void *sock_ptr) {
+    struct nw_socket *nw_socket = sock_ptr;
+    AWS_LOGF_DEBUG(AWS_LS_IO_SOCKET, "id=%p : start s_socket_internal_destroy", (void *)sock_ptr);
+
+    if (s_validate_event_loop(nw_socket->event_loop)) {
+        struct socket_close_complete_args *args =
+            aws_mem_calloc(nw_socket->allocator, 1, sizeof(struct socket_close_complete_args));
+
+        args->shutdown_complete_fn = nw_socket->on_socket_close_complete_fn;
+        args->user_data = nw_socket->close_user_data;
+        args->allocator = nw_socket->allocator;
+        args->nw_socket = nw_socket;
+        // At this point the internal ref count has been dropped to 0, and we are about to release the external ref
+        // count.
+        // However, we would still keep the external ref count alive until the s_close_complete_callback callback is
+        // invoked. Acquire another external ref count to keep the socket alive. It will be released in
+        // s_close_complete_callback.
+        aws_ref_count_acquire(&nw_socket->nw_socket_ref_count);
+        aws_task_init(&args->task, s_close_complete_callback, args, "SocketShutdownCompleteTask");
+
+        aws_event_loop_schedule_task_now(nw_socket->event_loop, &args->task);
+    } else {
+        // If we are not on the event loop
+        if (nw_socket->on_socket_close_complete_fn) {
+            nw_socket->on_socket_close_complete_fn(nw_socket->close_user_data);
+        }
+    }
+    s_release_event_loop(nw_socket);
+    aws_ref_count_release(&nw_socket->nw_socket_ref_count);
+}
+
+int aws_socket_init_apple_nw_socket(
+    struct aws_socket *socket,
+    struct aws_allocator *alloc,
+    const struct aws_socket_options *options) {
+    AWS_FATAL_ASSERT(options);
+    AWS_ZERO_STRUCT(*socket);
+
+    // Network Interface is not supported with Apple Network Framework yet
+    if (options->network_interface_name[0] != 0) {
+        AWS_LOGF_ERROR(
+            AWS_LS_IO_SOCKET,
+            "id=%p fd=%d: network_interface_name is not supported on this platform.",
+            (void *)socket,
+            socket->io_handle.data.fd);
+        return aws_raise_error(AWS_ERROR_PLATFORM_NOT_SUPPORTED);
+    }
+
+    struct nw_socket *nw_socket = aws_mem_calloc(alloc, 1, sizeof(struct nw_socket));
+    nw_socket->allocator = alloc;
+
+    socket->allocator = alloc;
+    socket->options = *options;
+    socket->impl = nw_socket;
+    socket->vtable = &s_vtable;
+
+    if (s_setup_socket_params(nw_socket, options)) {
+        aws_mem_release(alloc, nw_socket);
+        return AWS_OP_ERR;
+    }
+
+    aws_mutex_init(&nw_socket->synced_data.lock);
+    aws_mutex_init(&nw_socket->base_socket_synced_data.lock);
+    nw_socket->base_socket_synced_data.base_socket = socket;
+
+    s_set_socket_state(nw_socket, socket, INIT);
```

Comment:
Since this is init(), we can safely set the socket state without calling this function.